presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushRemoteExchangeThroughGroupId.java

```diff
@@ -0,0 +1,192 @@
+    private static final Capture<GroupIdNode> GROUP_ID = newCapture();
+    private static final Pattern<ExchangeNode> PATTERN = exchange()
+            .matching(exchange -> exchange.getScope().isRemote())
+            .matching(exchange -> exchange.getType() == REPARTITION)
+            .with(source().matching(
+                    groupId()
+                            .capturedAs(GROUP_ID)
+                            .matching(groupId -> !groupId.getCommonGroupingColumns().isEmpty())));
+
+    public PushRemoteExchangeThroughGroupId(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Pattern<ExchangeNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return shouldPushRemoteExchangeThroughGroupId(session);
+    }
+
+    @Override
+    public Result apply(ExchangeNode node, Captures captures, Context context)
+    {
+        GroupIdNode groupIdNode = captures.get(GROUP_ID);
+
+        List<VariableReferenceExpression> inputs = getOnlyElement(node.getInputs());
+        inputs = removeVariable(inputs, groupIdNode.getGroupIdVariable());
+        inputs = replaceAlias(inputs, groupIdNode.getGroupingColumns());
+
+        PartitioningScheme partitioningScheme = node.getPartitioningScheme();
+        List<VariableReferenceExpression> outputLayout = partitioningScheme.getOutputLayout();
+        outputLayout = removeVariable(outputLayout, groupIdNode.getGroupIdVariable());
+        outputLayout = replaceAlias(outputLayout, groupIdNode.getGroupingColumns());
+
+        Set<VariableReferenceExpression> commonGroupingColumns = groupIdNode.getCommonGroupingColumns();
+        List<VariableReferenceExpression> partitionColumns = replaceAlias(commonGroupingColumns, groupIdNode.getGroupingColumns());
+
+        // Check that new |partitionColumns| must be subset of original partition columns.
+        Map<VariableReferenceExpression, VariableReferenceExpression> groupingColumns = groupIdNode.getGroupingColumns();
+        List<VariableReferenceExpression> originalPartitionColumns =
+                partitioningScheme.getPartitioning().getVariableReferences()
+                        .stream()
+                        .map(expr -> groupingColumns.getOrDefault(expr, expr))
+                        .collect(toImmutableList());
+        if (!originalPartitionColumns.containsAll(partitionColumns)) {
+            return Result.empty();
+        }
+
+        // Create new PartitioningHandle.
+        PartitioningHandle partitioningHandle;
+        try {
+            partitioningHandle = createPartitioningHandle(context.getSession(), partitionColumns);
+        }
+        catch (PrestoException ex) {
+            // Some connectors do not support custom partitioning. Reuse existing handle as a fallback.
+            partitioningHandle = partitioningScheme.getPartitioning().getHandle();
```

Comment:
I would rather follow the logic in AddExchanges for explicitly checking for the system connector.  This should be the more common case (partitioning providers are only used for Large batch mode, and any non-system connector that's used as a partitioning provider will need to support createPartitioningHandle()), so it's better not to rely on exception handling for the main query path.

Reply:
Sounds good, let me update this!